src/imputors/knn.jl

```diff
@@ -4,83 +4,75 @@
-    num_nn::Int
+    k::Int
     dist::M
-    missing_neighbors_threshold::AbstractFloat
+    threshold::AbstractFloat
     context::AbstractContext
 end
 
-function KNN(; num_nn=1, missing_neighbors_threshold=0.5,
+function KNN(; k=1, threshold=0.5,
     dist=Euclidean(), context=Context())
 
-    if num_nn < 1
-        throw(ArgumentError("The number of nearset neighbors should be greater than 0"))
-    end
+    k < 1 && throw(ArgumentError("The number of nearset neighbors should be greater than 0"))
+
+    !(0 < threshold < 1) && throw(ArgumentError("Missing neighbors threshold should be within 0 to 1"))
 
     # to exclude missing value itself
-    KNN(num_nn + 1, dist, missing_neighbors_threshold, context)
+    KNN(k + 1, dist, threshold, context)
 end
 
 function impute!(data::AbstractMatrix{<:Union{T, Missing}}, imp::KNN) where T<:Real
 
     imp.context() do ctx
-        # transpose to D x N for KDTree
-        transposed = float.(collect(transpose(data)))
-
-        # Get our before and after views of our missing and non-missing data
-        mmask = ismissing.(transposed)
-        omask = .!mmask
-
-        mdata = transposed[mmask]
-        odata = transposed[omask]
+        # Get mask array first (order of )
+        mmask = ismissing.(transpose(data))
 
-        # fill missing value as mean first
-        impute!(transposed, Fill(; value=mean, context=ctx))
+        # fill missing value as mean value
+        impute!(data, Fill(; value=mean, context=ctx))
 
-        # disallow missing to compute kdtree
-        transposed = disallowmissing(transposed)
+        # then, transpose to D x N for KDTree
+        transposed = transpose(disallowmissing(data))
 
         kdtree = KDTree(transposed, imp.dist)
-        idxs, dists = NearestNeighbors.knn(kdtree, transposed, imp.num_nn, true)
+        idxs, dists = NearestNeighbors.knn(kdtree, transposed, imp.k, true)
 
         idxes = CartesianIndices(transposed)
 
         # iterate all value
         for I in CartesianIndices(transposed)
             if mmask[I] == 1
                 w = 1.0 ./ dists[I[2]]
-                #@show idxs[I[2]][2], dists[I[2]][2], sum(w[2:end])
-
+                ws = sum(w[2:end])
                 missing_neighbors = ismissing.(transposed[:, idxs[I[2]]][:, 2:end])
+
                 # exclude missing value itself because distance would be zero
-                if isnan(sum(w[2:end])) || isinf(sum(w[2:end])) || sum(w[2:end]) == 0.0
-                    # if distance is zero, keep mean imputation
+                if isnan(ws) || isinf(ws) || iszero(ws)
+                    # if distance is zero or not a number, keep mean imputation
                     transposed[I] = transposed[I]
-                elseif count(x -> !=(0, x), mapslices(sum, missing_neighbors, dims=[1])) >
-                    imp.num_nn * imp.missing_neighbors_threshold
+                elseif count(!iszero, mapslices(sum, missing_neighbors, dims=1)) >
+                    imp.k * imp.threshold
```

Comment:
Again, you could probably just save this to a variable earlier on, rather than recomputing it each time.